### PR TITLE
Handy Uri Template

### PR DESCRIPTION
**Pull request by casz**
Just initial work, to get some feedback

cc @stephenc

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -120,6 +120,11 @@
       <scope>test</scope>
     </dependency>
     <dependency>
+      <groupId>com.damnhandy</groupId>
+      <artifactId>handy-uri-templates</artifactId>
+      <version>2.1.6</version>
```
Comment (stephenc) [inline pom.xml]: We’ll need to either pull this into a library plugin or use maskClasses to shield ourselves from class conflicts

**Comment by casz** [comment]
I'll have the next bit up later today

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java [hunk 2]
```diff
@@ -386,7 +419,12 @@ public void removeCommitWebHook(@NonNull BitbucketWebHook hook) throws IOExcepti
         if (StringUtils.isBlank(hook.getUuid())) {
             throw new BitbucketException("Hook UUID required");
         }
-        deleteRequest(V2_API_BASE_URL + owner + "/" + repositoryName + "/hooks/" + Util.rawEncode(hook.getUuid()));
+        String url = UriTemplate.fromTemplate(REPO_URL_TEMPLATE + "/hooks/{uuid}")
+                .set("owner", owner)
+                .set("repo", repositoryName)
+                .set("uuid", Util.rawEncode(hook.getUuid()))
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: Don't think i need encoding here as Handy Uri should handle it

Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: Don't think i need encoding here as Handy Uri should handle it

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java [hunk 3]
```diff
@@ -683,14 +743,23 @@ private String deleteRequest(String path) throws IOException {
...
+                .set("limit", 500);
+        String url = template.expand();
+        String response = getRequest(url);
         Map<String,Object> content = JsonParser.mapper.readValue(response, new TypeReference<Map<String,Object>>(){});
         Map page = (Map) content.get("children");
         List<Map> values = (List<Map>) page.get("values");
         collectFileAndDirectories(parent, path, values, files);
         while (!(boolean)page.get("isLastPage")){
-            response = getRequest(url + String.format("&start=%s&limit=%s", start, 500));
+            start += (int) content.get("size");
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java]: We were never counting up the pages 😕

Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java]: We were never counting up the pages 😕 
Here we are 
https://github.com/casz/bitbucket-branch-source-plugin/blob/0f6a1060a67598df896406b11d15587c0541bf4e/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java#L725-L729

However not here
https://github.com/casz/bitbucket-branch-source-plugin/blob/0f6a1060a67598df896406b11d15587c0541bf4e/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java#L691-L695

Comment (stephenc) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java]: Can you create a JIRA and assign yourself (or let me know the JIRA and i’ll Assign it to you)

Good catch

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java (hunk 2 above)
Comment (stephenc) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: Yep, remove the rawEncode as it’s `{uuid}`. You’d need to use raw encode if the template was `{+uuid}`

**Comment by Hellspam** [comment]
Can you maybe make a larger limit for the getBranches function while you are at it? Something like 100 maybe instead of the current (25 I think by default).

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java (hunk 3 above)
Comment (casz) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java]: Jira issue created: https://issues.jenkins-ci.org/browse/JENKINS-48279

File: pom.xml (hunk 1 above)
Comment (casz) [reply pom.xml]: Looking at the deps: https://github.com/damnhandy/Handy-URI-Templates/blob/master/pom.xml#L473-L533

Comment (casz) [reply pom.xml]: Looking at the deps: https://github.com/damnhandy/Handy-URI-Templates/blob/master/pom.xml#L473-L533
They are all optional or test scoped except for joda-time.
If we were to put into a library plugin, this is the way to go right: https://github.com/jenkinsci/apache-httpcomponents-client-4-api-plugin with some POM magic?

Comment (vivek) [reply pom.xml]: @casz yes, thats the way to go about library plugin. joda-time libraries are used at other places too. It might cause conflicts.

**Comment by casz** [comment]
I'll try and wrap it in a Jenkins plugin and put it up Jenkins repo tonight, and start the process of having it hosted 👍

File: pom.xml (hunk 1 above)
Comment (casz) [reply pom.xml]: @stephenc if possible, please advice if I should be doing something special against joda-time or any of the optional dependencies.

Comment (stephenc) [reply pom.xml]: It they are optional, then they will not be transitive.

I think joda-time may need its own api plugin, in which case you then have to add a *jenkins plugin dependency* to your api plugin... fun times

Comment (casz) [reply pom.xml]: @stephenc I give it a shot for the joda-time too :)

Comment (casz) [reply pom.xml]: Joda-time should be safe as well. However I will gladly wrap it in a plugin :) 
```xml
  <dependencies>
    <dependency>
      <groupId>org.joda</groupId>
      <artifactId>joda-convert</artifactId>
      <version>1.2</version>
      <scope>compile</scope>
      <optional>true</optional><!-- mandatory in Scala -->
    </dependency>
    <dependency>
      <groupId>junit</groupId>
      <artifactId>junit</artifactId>
      <version>3.8.2</version>
      <scope>test</scope>
    </dependency>
  </dependencies>
```

Comment (casz) [reply pom.xml]: Currently waiting for approval: https://issues.jenkins-ci.org/projects/HOSTING/issues/HOSTING-476

**Comment by casz** [comment]
@stephenc rebased onto master and using snapshot while I wait for hosting 👍 Seems like we are getting rid of all the ugly string format and encoding. Could you please take a look, if I missed anything.

If you still think Yoda, needs to be wrapped. All the dependencies are optional or as test scope.

File: pom.xml [hunk 4]
```diff
@@ -135,6 +135,11 @@
       <classifier>tests</classifier>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.jenkins-ci.plugins</groupId>
+      <artifactId>handy-uri-templates-2-api</artifactId>
+      <version>2.1.6-1.0-20180102.164938-1</version>
```
Comment (stephenc) [inline pom.xml]: Merging blocked until this is released

Comment (casz) [reply pom.xml]: it just got hosted, so I'll upload it tonight 👍

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java [hunk 5]
```diff
@@ -460,13 +505,17 @@ private BitbucketRepositoryHooks parsePaginatedRepositoryHooks(String response)
     @Override
     @CheckForNull
     public BitbucketTeam getTeam() throws IOException, InterruptedException {
+        String url = UriTemplate.fromTemplate("{+base}{/owner}")
+                .set("base", V2_TEAMS_API_BASE_URL)
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: perhaps we should change this template, I don't like the `set "base"`
WDYT?

File: src/test/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/BitbucketCloudEndpointTest.java [hunk 6]
```diff
@@ -39,8 +43,32 @@ public void smokes() {
 
     @Test
     public void getRepositoryUrl() {
+
```
Comment (casz) [inline src/test/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/BitbucketCloudEndpointTest.java]: 👻 whitespace

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java [hunk 7]
```diff
@@ -726,13 +827,24 @@ public InputStream getFileContent(BitbucketSCMFile file) throws IOException, Int
         String path = encodePath(file.getPath());
         String ref = URLEncoder.encode(file.getRef(), "UTF-8");
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java]: This is no longer needed?

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java [hunk 8]
```diff
@@ -726,13 +827,24 @@ public InputStream getFileContent(BitbucketSCMFile file) throws IOException, Int
         String path = encodePath(file.getPath());
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java]: This is no longer needed?

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java [hunk 9]
```diff
@@ -689,15 +779,26 @@ private String deleteRequest(String path) throws IOException {
         String path = encodePath(directory.getPath());
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java]: This is no longer needed?

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java [hunk 10]
```diff
@@ -689,15 +779,26 @@ private String deleteRequest(String path) throws IOException {
...
         String ref = URLEncoder.encode(directory.getRef(), "UTF-8");
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java]: This is no longer needed?

**Comment by casz** [comment]
Tested with both bitbucket cloud and bitbucket server that URL's and scans still work 👍 🚀

**Comment by stephenc** [comment]
Can we get uri templates on https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/fed749a4860d0171d89422fb0f5aa8dcfc48c6cd/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java#L1040-L1069 too

**Comment by casz** [comment]
@stephenc it's there :)

https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/80/files#diff-dbebe1606f4711ed03640cf707dc6c23
https://github.com/casz/bitbucket-branch-source-plugin/blob/34374d3f9a625d768bd09c2974e5ec0f8d607777/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java#L1040-L1092

I guess it could be made even BETTER 👍

**Comment by casz** [comment]
@stephenc slightly improved version https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/80/commits/803dbe4c1ffeb53b5dac929e4315aff46fb910d5

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java [hunk 11]
```diff
@@ -1036,52 +1037,45 @@ public DescriptorImpl getDescriptor() {
...
+                        .fromTemplate(getServerUrl() + SERVER_REPO_TEMPLATE + "/compare/commits{?sourceBranch}")
                         .set("owner", repoOwner)
                         .set("repo", repository)
                         .set("sourceBranch", Constants.R_HEADS + head.getName());
-                title = null;
             }
-            result.add(new BitbucketLink("icon-bitbucket-branch", template.expand()));
-            result.add(new ObjectMetadataAction(title, null, template.expand()));
         }
+        if (head instanceof PullRequestSCMHead) {
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java]: The if-else above is now only for URL template and this is now deduped.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java [hunk 12]
```diff
@@ -1036,52 +1037,45 @@ public DescriptorImpl getDescriptor() {
...
+            PullRequestSCMHead pr = (PullRequestSCMHead) head;
+            title = getPullRequestTitleCache().get(pr.getId());
+            ContributorMetadataAction contributor = getPullRequestContributorCache().get(pr.getId());
+            if (contributor != null) {
+                result.add(contributor);
+            }
+        }
+        String url = template.expand();
+        result.add(new BitbucketLink("icon-bitbucket-branch", url));
+        result.add(new ObjectMetadataAction(title, null, url));
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java]: More deduping

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java [hunk 13]
```diff
@@ -1014,17 +1016,16 @@ public DescriptorImpl getDescriptor() {
...
-            result.add(new BitbucketLink("icon-bitbucket-repo",
-                    getServerUrl() + "/projects/" + repoOwner + "/repos/" + repository));
-            result.add(new ObjectMetadataAction(r.getRepositoryName(), null,
-                    getServerUrl() + "/projects/" + repoOwner + "/repos/" + repository));
+            template = UriTemplate.fromTemplate(getServerUrl() + SERVER_REPO_TEMPLATE);
         }
+        template.set("owner", repoOwner).set("repo", repository);
+        String url = template.expand();
+        result.add(new BitbucketLink("icon-bitbucket-repo", url));
+        result.add(new ObjectMetadataAction(r.getRepositoryName(), null, url));
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java]: Some much needed deduping

**Comment by vivek** [comment]
Looks like its ready to be merged. @stephenc, looks good to you?

**Comment by stephenc** [comment]
Let’s go

**Pull request merged**